File: migrator/partition.py
from datetime import datetime, timezone
from enum import Enum


class Partitioning(str, Enum):
    MINUTES = "MINUTES"
    HOURS = "HOURS"
    DAYS = "DAYS"
    MONTHS = "MONTHS"
    YEARS = "YEARS"
    INDEFINITE = "INDEFINITE"


def compute_partition(ts_ms: int, strategy: Partitioning) -> int:
    if strategy == Partitioning.INDEFINITE:
        return 0
    dt = datetime.fromtimestamp(ts_ms / 1000, tz=timezone.utc)
    if strategy == Partitioning.MINUTES:
        t = dt.replace(second=0, microsecond=0)
    elif strategy == Partitioning.HOURS:
        t = dt.replace(minute=0, second=0, microsecond=0)
    elif strategy == Partitioning.DAYS:
        t = dt.replace(hour=0, minute=0, second=0, microsecond=0)
    elif strategy == Partitioning.MONTHS:
        t = dt.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    elif strategy == Partitioning.YEARS:
        t = dt.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
    return int(t.timestamp() * 1000)

File: Partition.cs
namespace TbMigrator;

public static class Partition
{
    public static long Compute(long tsMs, string strategy)
    {
        var dt = DateTimeOffset.FromUnixTimeMilliseconds(tsMs).UtcDateTime;
        DateTimeOffset t = strategy switch
        {
            "MINUTES"    => new DateTimeOffset(dt.Year, dt.Month, dt.Day, dt.Hour, dt.Minute, 0, TimeSpan.Zero),
            "HOURS"      => new DateTimeOffset(dt.Year, dt.Month, dt.Day, dt.Hour, 0, 0, TimeSpan.Zero),
            "DAYS"       => new DateTimeOffset(dt.Year, dt.Month, dt.Day, 0, 0, 0, TimeSpan.Zero),
            "MONTHS"     => new DateTimeOffset(dt.Year, dt.Month, 1, 0, 0, 0, TimeSpan.Zero),
            "YEARS"      => new DateTimeOffset(dt.Year, 1, 1, 0, 0, 0, TimeSpan.Zero),
            "INDEFINITE" => DateTimeOffset.MinValue,
            _ => throw new ArgumentException($"Unknown partitioning: {strategy}")
        };
        return strategy == "INDEFINITE" ? 0 : t.ToUnixTimeMilliseconds();
    }
}
